moped-etl/arcgis/components_to_agol.py

```diff
@@ -115,17 +129,81 @@ def make_all_features(data):
                 attributes=component,
             )
             all_features["combined"].append(line_feature)
+
+            project_component_id = feature["attributes"]["project_component_id"]
+
+            # Find the elements in exploded_geometry which match the project_component_id
+            matching_elements = [
+                element
+                for element in exploded_geometry
+                if element["project_component_id"] == project_component_id
+            ]
```

Comment:
Yea, that's a good thought for sure, because, at the very least, it'd save bandwidth not pulling down duplicates of the data, just grabbing the geometry and a primary key. 

Since this is an ETL though, I think it's less of a concern as it should be if this were in a reactive application, but I could easily be convinced that it's worth it. @mddilley, do you have direction on if you'd like to see that optimization that John suggested?

Reply:
One advantage that I could see here is having to only update one of the queries with a field when it comes time to add a new one. Does this feel like a big lift?

Reply:
Frank and I chatted about John's feedback in combination with [this comment](https://github.com/cityofaustin/atd-moped/pull/1397#discussion_r1755491619). Just to make sure we are all talking about the same thing - i was thinking something like:
```
query GetProjectsComponents($where: component_arcgis_online_view_bool_exp!) {
  component_arcgis_online_view(where: $where) {
    <all of the fields>
  }
  exploded_component_arcgis_online_view(where: $where) {
    <fields needed for the exploded part>
  }
}
```

Reply:
@frankhereford @johnclary mentioning y'all because i realized this thread is pretty buried!

Reply:
💯 That's what I had originally written out and was trying to go with, but then I got hung up on the line:

```
query GetProjectsComponents($where: component_arcgis_online_view_bool_exp!) {
```

I'm unsure what is going on here, and how to expand it to cover both views that we're selecting out of. The `$` in front of the where and this unknown value -- I'm not even sure what to call it. If anyone has some docs they can point me to, I would be massively grateful. I am a graphql newb, for sure.

Reply:
ahhh gotcha! i see now. The Hasura docs are thin on this, but it is a way to dynamically set the filter conditions. I'll try to find an example or doc somewhere, but the secret sauce is this part of the script (line 168) where we set the filter conditions based on the value of `args.full`.

```python
variables = (
        {"where": {}} # no filter w/ full replace
        if args.full
        else {"where": {"project_updated_at": {"_gt": args.date}}} # date filter w/ incremental replace
    )
```
I see the issue because we are dealing with two different queries, component and exploded, which require different variable types, `component_arcgis_online_view_bool_exp!` and `exploded_component_arcgis_online_view!`. So it would have to be more like:
```
query GetProjectsComponents($where: component_arcgis_online_view_bool_exp!, $whereExploded: exploded_component_arcgis_online_view_bool_exp!) {
  component_arcgis_online_view(where: $where) {
    <all of the fields>
  }
  exploded_component_arcgis_online_view(where: $whereExploded) {
    <fields needed for the exploded part>
  }
}
```
and then an update to the vars passed in `make_hasura_request`. Going to think about this and come back to it in the morning.

Reply:
fwiw, you also have the option of using python string formatting to compose the query / make it reusable.

 but my comment was referring not to the Hasura query, but the actual construction of this view. We have the option of slimming down the new exploded geometry view so that it only contains geometry data, and not any component attribute columns.

in other words, you could change lines 142 to 150 to look like this:

```python
            for record in matching_exploded_geometry_records:
                geometry = record.pop("geometry")
                esri_geometry_key = "point"

                feature = make_esri_feature(
                    esri_geometry_key="point",
                    geometry=geometry,
                    attributes=component,   # use the component properties from the matching component
                )
```

that would have the benefit of having fewer places where we need to maintain the code for this ETL.

Reply:
thanks John for the insight here and Frank for implementing. The latest iteration of the exploded view makes this very maintainable for next time around. 🙏